// @hashgraph/sdk/lib/GetByKeyQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const QueryBuilder_1 = require("./QueryBuilder");
const QueryHeader_pb_1 = require("./generated/QueryHeader_pb");
const FileId_1 = require("./file/FileId");
const ContractId_1 = require("./contract/ContractId");
const AccountId_1 = require("./account/AccountId");
const pb = require("./generated/GetByKey_pb");
class GetByKeyQuery extends QueryBuilder_1.QueryBuilder {
    constructor() {
        const header = new QueryHeader_pb_1.QueryHeader();
        super(header);
        this._builder = new pb.GetByKeyQuery();
        this._builder.setHeader(header);
        this._inner.setGetbykey(this._builder);
    }
    setKey(publicKey) {
        this._builder.setKey(publicKey._toProtoKey());
        return this;
    }
    _doValidate( /* errors: string[] */) {
        // Do nothing
    }
    get _method() {
        throw new Error("not implemented; no service in Hedera exposes this method");
    }
    _mapResponse(response) {
        return response.getGetbykey().getEntitiesList().filter((id) => id.hasContractid() ||
            id.hasAccountid() || id.hasFileid()).map((id) => {
            if (id.hasAccountid()) {
                return {
                    type: "ACCOUNT",
                    accountId: AccountId_1.AccountId.fromProto(id.getAccountid())
                };
            }
            else if (id.hasContractid()) {
                return {
                    type: "CONTRACT",
                    contractId: ContractId_1.ContractId.fromProto(id.getContractid())
                };
            }
            else if (id.hasFileid()) {
                return {
                    type: "FILE",
                    fileId: FileId_1.FileId.fromProto(id.getFileid())
                };
            }
            throw new Error("unreachable");
        });
    }
}
exports.GetByKeyQuery = GetByKeyQuery;
